### ISSUE TITLE
Remove Guava as a dependency

### ISSUE THREAD

**Issue by jasontedor** [elastic/elasticsearch#13224]
This is a list of the imports from Guava that can be found in core Elasticsearch today. Let's check these off as we go and reference the corresponding issues and pull requests. Most of these will be relatively straightforward with the most serious ones being replacing the caching layer. Those will probably require full issues and careful review cycles.

- [x] remove `import com.google.common.annotations.Beta;` fe46227db871b6d1af6333c21f4bc684d350dd97
- [x] remove `import com.google.common.base.Charsets;` #13533
- [x] remove `import com.google.common.base.Function;` #13533
- [x] remove `import com.google.common.base.Joiner;` #13572 
- [x] remove `import com.google.common.base.Objects;` #13355 
- [x] remove `import com.google.common.base.Preconditions;` #13540 
- [x] remove `import com.google.common.base.Predicate;` #13349 
- [x] remove `import com.google.common.base.Predicates;` #13349 
- [x] remove `import com.google.common.base.Strings;` a6ffe8f6d50e207b44c327c99cb304f20bb07027
- [x] remove `import com.google.common.base.Throwables;` #13409 
- [x] remove `import com.google.common.cache.*;` #13717, #13879  
- [x] remove `import com.google.common.cache.Cache;` #13717, #13879 
- [x] remove `import com.google.common.cache.CacheBuilder;` #13717, #13879 
- [x] remove `import com.google.common.cache.CacheLoader;` 30bd46ab2534e4366eb840e0b44b04fe4713bd19
- [x] remove `import com.google.common.cache.LoadingCache;` 30bd46ab2534e4366eb840e0b44b04fe4713bd19
- [x] remove `import com.google.common.cache.RemovalListener;` #13717, #13879 
- [x] remove `import com.google.common.cache.RemovalNotification;` #13717, #13879 
- [x] remove `import com.google.common.cache.Weigher;` #13717, #13879 
- [x] remove `import com.google.common.collect.*;`
- [x] remove `import com.google.common.collect.ArrayListMultimap;` #13562, 257f3e862f5cd0cd24777efe51bfae6d1d03fedf
- [x] remove `import com.google.common.collect.Collections2;` #13533
- [x] remove `import com.google.common.collect.FluentIterable;` #13349, 257f3e862f5cd0cd24777efe51bfae6d1d03fedf
- [x] remove `import com.google.common.collect.HashMultimap;` #13562, 257f3e862f5cd0cd24777efe51bfae6d1d03fedf 
- [x] remove `import com.google.common.collect.EvictingQueue;` #13903
- [x] remove `import com.google.common.collect.ImmutableCollection;` #13909
- [x] remove `import com.google.common.collect.ImmutableList;` #13227 
- [x] remove `import com.google.common.collect.ImmutableMap.Builder;` #13939 
- [x] remove `import com.google.common.collect.ImmutableMap;` #13762, #13939
- [x] remove `import com.google.common.collect.ImmutableSet;` #13724, #13754
- [x] remove `import com.google.common.collect.ImmutableSortedMap;` #13525 
- [x] remove `import com.google.common.collect.ImmutableSortedSet;` 65353b8a3289d942b391fd71bbf4ef4b693dfea5 
- [x] remove `import com.google.common.collect.Iterators;` #13916
- [x] remove `import com.google.common.collect.ForwardingSet;` #13720
- [x] remove `import com.google.common.collect.Iterables;` #13559
- [x] remove `import com.google.common.collect.Lists;` #13170 
- [x] remove `import com.google.common.collect.Maps;` #13438 
- [x] remove `import com.google.common.collect.Multimap;` #13562
- [x] remove `import com.google.common.collect.MultimapBuilder;` #13562
- [x] remove `import com.google.common.collect.ObjectArrays;` #13562
- [x] remove `import com.google.common.collect.Queues;` #13498 
- [x] remove `import com.google.common.collect.Sets;` #13463
- [x] remove `import com.google.common.collect.UnmodifiableIterator;` #13562
- [x] remove `import com.google.common.hash.HashCode;` #13907
- [x] remove `import com.google.common.hash.HashFunction;` #13907
- [x] remove `import com.google.common.hash.Hashing;` #13907
- [x] remove `import com.google.common.io.ByteStreams;` #13562
- [x] remove `import com.google.common.io.Files;` #13562, 257f3e862f5cd0cd24777efe51bfae6d1d03fedf
- [x] remove `import com.google.common.io.Resources;` #13908
- [x] remove `import com.google.common.jimfs.Configuration;` remain as test dependency only
- [x] remove `import com.google.common.jimfs.Jimfs;` remain as test dependency only
- [x] remove `import com.google.common.math.LongMath;` #13575 
- [x] remove `import com.google.common.net.InetAddresses;` #13905
- [x] remove `import com.google.common.primitives.Ints;` #13596
- [x] remove `import com.google.common.primitives.Longs;` #13562
- [x] remove `import com.google.common.util.concurrent.AtomicLongMap;` #13562
- [x] remove `import com.google.common.util.concurrent.Futures;` #13524
- [x] remove `import com.google.common.util.concurrent.ListenableFuture;` #13524
- [x] remove `import com.google.common.util.concurrent.MoreExecutors;` #13524
- [x] remove `import com.google.common.util.concurrent.SettableFuture;` #13524
- [x] remove `import com.google.common.util.concurrent.UncheckedExecutionException;` #13562 
- [x] remove `import static com.google.common.base.Preconditions.*;` #13540 
- [x] remove `import static com.google.common.base.Preconditions.checkArgument;` #13540 
- [x] remove `import static com.google.common.base.Preconditions.checkNotNull;` #13493 
- [x] remove `import static com.google.common.base.Preconditions.checkState;` #13540 
- [x] remove `import static com.google.common.base.Predicates.isNull;` #13349 
- [x] remove `import static com.google.common.collect.Lists.*;` #13170 
- [x] remove `import static com.google.common.collect.Lists.newArrayList;` #13170 
- [x] remove `import static com.google.common.collect.Lists.newArrayListWithCapacity;` #13170
- [x] remove `import static com.google.common.collect.Maps.filterEntries;` #13438 
- [x] remove `import static com.google.common.collect.Maps.newHashMap;` #13438 
- [x] remove `import static com.google.common.collect.Maps.newHashMapWithExpectedSize;` #13438 
- [x] remove `import static com.google.common.collect.Sets.newHashSet;` #13463 

Note that `jimfs` is in an artifact separate from `guava`, but depends on `guava` itself; it will remain as a test-only dependency.

The above was generated via:

    ack --type java -h "import\s+(static\s+)?com\.google\.common" | sort | uniq | perl -p -e 's/^(.*)/- [ ] remove `$1`/g'

**Comment by nik9000** [comment]
I think some of these may be worth importing directly - `Splitter` and `Joiner` come to mind. They are cute and convenient. We should just be careful not to drift from Guava's source or upgrades will be difficult.

I think other of these - `newHashMap`, `newHashSet`, `filtering`, etc, are worth banning and removing entirely because they just aren't that useful. The `new*` methods in particular are pretty much obsolete with the diamond operator.

I guess what I'm getting at is that we should handle each of these in a case by case basic.

**Comment by jasontedor** [comment]
Removing the v2.1.0 label; I suspect that some of these will stick around for convenience until we migrate to Java 8 and can replace uses of things like `Iterables.filter/Predicate` with `Stream.filter/native Predicate`.

**Comment by xuzha** [comment]
Just came across to google.io tonight, not sure if @jasontedor or others has been working on it.

**Comment by nik9000** [comment]
Can we talk about `com.google.common.base.Preconditions`? I kind of like it for times when you don't want an `assert`. I find its generally less clutter-y than lots of `if`s.

**Comment by rjernst** [comment]
Do we need more than `checkArgument`? We can replace all the `checkNotNull` with `Objects.requireNonNull`. And `checkArgument` is a super simple method...doesn't seem worthwhile to keep an entire complex library around just for that.

**Comment by jasontedor** [comment]
@nik9000:

> Can we talk about com.google.common.base.Preconditions? I kind of like it for times when you don't want an assert. I find its generally less clutter-y than lots of ifs.

Have you seen the comment regarding the use of these method in the [Javadocs](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Preconditions.html#checkArgument(boolean)), specifically:

> Remember that parameter values for message construction must all be computed eagerly, and autoboxing and varargs array creation may happen as well, even when the precondition check then succeeds (as it should almost always do in production). In some circumstances these wasted CPU cycles and allocations can add up to a real problem. Performance-sensitive precondition checks can always be converted to the customary form

Reducing object churn in Elasticsearch is immensely important because a garbage collection cycle is no different than a network partition from a failure perspective. Increasing the chance of long runtime latencies from excessive garbage collection cycles is detrimental to cluster health.

**Comment by jasontedor** [comment]
@rjernst:

> And `checkArgument` is a super simple method...doesn't seem worthwhile to keep an entire complex library around just for that.

+1

**Comment by jasontedor** [comment]
This should not be taken as a judgement on Guava. Guava is a popular library, and for good reasons. But the rationale here is rather simple:

* A lot of the benefits of Guava have been supplanted by additions to the JDK in JDK 7 and JDK 8 (e.g., the diamond operator, lambdas, and streams).
* We are no longer shading dependencies in the Elasticsearch jar (see #13252).
* We do not have a proper Java client today; this means that developers that depend on the Elasticsearch jar are likely going to have to deal with Guava version conflicts (which are likely since Guava is popular).
* Developers do embed Elasticsearch inside web containers which creates conflicts with the provided versions of Guava.
* Plugin developers encounter issues with Guava version conflicts (again, because Guava is popular.
* All else equal, fewer dependencies is a positive.

In short, the internal benefits have been reduced, and there are external benefits to developers to removing Guava.

**Comment by nik9000** [comment]
> And checkArgument is a super simple method...doesn't seem worthwhile to keep an entire complex library around just for that.

I wasn't clear - I'm not advocating that. I'm advocating that we consider importing Preconditions, or at least portions of it rather than rewriting them all at the call sight.


> Reducing object churn in Elasticsearch is immensely important because a garbage collection cycle is no different than a network partition from a failure perspective. Increasing the chance of long runtime latencies from excessive garbage collection cycles is detrimental to cluster health.

I like this argument much better. To me it goes like this:
1. We want to reduce our object allocation rate.
2. `checkArgument` and its ilk always attempt to make your code more readable at the cost of potentially more allocations.
3. This is often a fine tradeoff but for Elasticsearch its never what we want because of point 1.
4. Its worse because the places where it is a fine tradeoff aren't always obvious in Elasticsearch. Put another way, it can be difficult to tell when you are in a commonly called code path. So its never worth the risk for the small readability payoff.

I promise I won't freak out about all the parts of Guava. Its clearly fine to replace `Charsets.UTF_8` with `StandardCharsets.UTF_8` for example.

**Comment by jasontedor** [comment]
@nik9000:

> I like this argument much better. To me it goes like this:
1. We want to reduce our object allocation rate.
2. `checkArgument` and its ilk always attempt to make your code more readable at the cost of potentially more allocations.
3. This is often a fine tradeoff but for Elasticsearch its never what we want because of point 1.
4. Its worse because the places where it is a fine tradeoff aren't always obvious in Elasticsearch. Put another way, it can be difficult to tell when you are in a commonly called code path. So its never worth the risk for the small readability payoff.

Exactly. Tangentially, another place to be aware of this is calls to `logger.(debug|error|info|trace|warn)`; any invocations of these that are allocating strings (e.g., via concatenation or calls to `Object#toString`) to build the method parameters whether or not the corresponding log level is turned contribute to object churn. In cases where there are allocations occurring, it's best to wrap the calls in the corresponding check against the `logger.is*Enabled` method. I've seen exactly this problem wreak havoc on HBase heaps.

**Comment by nik9000** [comment]
I squashed the incomplete items together in the list to make the remaining work easier to read.

**Comment by nik9000** [comment]
It looks like the two remaining ones are claimed by me. I got a good start to them on the plane and will push something reviewable early next week.
